### PR TITLE
Bugfix/2785 : Backend fix for Idea Name Cannot Be Reused After Deletion

### PR DESCRIPTION
**Pull request by royshunhung**
I have successfully completed the bug fix for ISSUE https://github.com/dev-launchers/dev-launchers-platform/issues/2785.
Please also review front end PR https://github.com/dev-launchers/dev-launchers-platform/pull/2829 related to this bug.

What's Included:

Added Logic to filter out deleted idea_name from API call in lifecycles.js
Added partial unique index under migrator/ensure-partial-index.js
Removed unique attribute from idea_name for Idea_card under idea-card/schema.json & types/generated/contentTypes.d.ts

Files Affected:

/lifecycles.js
/migrator/ensure-partial-index.js
/idea-card/schema.json
/types/generated/contentTypes.d.ts

Technical Notes:

Previous name taken logic on front end does not check if the previous idea name was deleted already, I added the logic to account for this during API call. I also added a file to force partial unique index for idea_name for idea_card. The unique attribute from idea_name is removed accordingly due to the new partial unique constraint in place. The implementation is complete and ready for staging deployment. 

Testing Method:

Testing was done by verifying the existence of partial index in SQLite3 database. The index rule does exist within SQLite3 database. After removing the unique attribute for ideaName initially, I also tested that if this allows duplicate entry for ideaName. It allows for entry with two identical idea_name with workshopping status. Hence, I moved to check the behavior after the partial index. Afterward, strapi only allows duplicate ideaName if the previous idea_name has the status of deleted.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -26,6 +27,9 @@ module.exports = {
 
     await runMigrations(strapi);
 
+    // Ensure custom partial unique index exists after Strapi startup/migration logic.
+    await ensureIdeaNameActiveIndex(strapi);
+
```
[the remark this answers is not in the document]
Comment (royshunhung) [reply src/index.js:32]: PR request aligned. Mechanism is now under src/migrators

File: src/api/idea-card/content-types/idea-card/lifecycles.js [hunk 2]
```diff
@@ -8,7 +8,11 @@ module.exports = {
 
     const previousData = await strapi.db.query("api::idea-card.idea-card")
       .findOne({
-        filters: { ideaName: { $eqi: event.params.data.ideaName } },
+        filters: { 
+          ideaName: { $eqi: event.params.data.ideaName } ,
+          status: {$ne: "deleted"}
+      
+      },
       });
```
[the remark this answers is not in the document]
Comment (royshunhung) [reply src/api/idea-card/content-types/idea-card/lifecycles.js:17]: Confirmed no issue with functionality. Closing

File: README.md [hunk 3]
```diff
@@ -1,5 +1,7 @@
  # Dev Launchers Strapi Service
 
+
+
```
[the remark this answers is not in the document]
Comment (royshunhung) [reply README.md]: Extra line removed. Closing

File: src/api/idea-card/content-types/idea-card/lifecycles.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (royshunhung) [reply src/api/idea-card/content-types/idea-card/lifecycles.js:17]: Indentation fixed

[the remark this answers is not in the document]
Comment (royshunhung) [reply src/api/idea-card/content-types/idea-card/lifecycles.js:17]: Have measure in place for website to check for duplicate name

File: src/migrators/ensure-partial-index.js
Comment (chungthuang) [inline src/migrators/ensure-partial-index.js:1]: Usually migration files are numbered so we can know the order they are applied. https://docs-v4.strapi.io/dev-docs/database-migrations#creating-a-migration-file suggests using the creation time of the migration.

File: src/migrators/ensure-partial-index.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+
+async function ensureIdeaNameActiveIndex(strapi) {
+  const knex = strapi.db.connection;
+
+  await knex.raw(`
+    CREATE UNIQUE INDEX IF NOT EXISTS idx_idea_cards_idea_name_active
+    ON idea_cards (LOWER(idea_name))
+    WHERE status IS NULL OR status <> 'deleted'
```
Comment (chungthuang) [inline src/migrators/ensure-partial-index.js]: When can status be null?

File: src/index.js [hunk 5]
```diff
@@ -26,6 +27,9 @@ module.exports = {
 
     await runMigrations(strapi);
```
Comment (chungthuang) [inline src/index.js:28]: Based on the function name, I would assume it applies the `ensureIdeaNameActiveIndex` migration.

File: README.md [hunk 6]
```diff
@@ -1,5 +1,6 @@
  # Dev Launchers Strapi Service
 
+
```
Comment (chungthuang) [inline README.md]: Cam we remove this?

File: src/migrators/ensure-partial-index.js (hunk 4 above)
Comment (royshunhung) [reply src/migrators/ensure-partial-index.js]: Status can never be Null now that i think about it. I can remove the condition when status is NULL.

File: README.md (hunk 6 above)
Comment (royshunhung) [reply README.md]: Removed

File: src/index.js (hunk 5 above)
Comment (royshunhung) [reply src/index.js:28]: Currently, the logic of migration is not working correctly. I was able to verify that migration had run and index created. However, the newly index was removed post migration. Hence, I break the logic out to a separate line.

File: src/api/idea-card/content-types/idea-card/schema.json [hunk 7]
```diff
@@ -15,7 +15,7 @@
     "ideaName": {
       "type": "string",
       "required": true,
-      "unique": true
```
Comment (tanmaydhal) [inline src/api/idea-card/content-types/idea-card/schema.json:18]: Why the change here? Sorry I don;t have enough context on this ticket

Comment (royshunhung) [reply src/api/idea-card/content-types/idea-card/schema.json:18]: Hi, the change is for removing unique attribute for ideaName.

In this ticket, we want to allow for creation of the same ideaName if the previous IdeaCard with the same ideaName is "deleted". Hence we have to remove the unique attribute to allow for duplicate and put into another index rule to allow for this case.

File: src/migrators/ensure-partial-index.js
Comment (royshunhung) [reply src/migrators/ensure-partial-index.js:1]: Currently, the migration logic doesn't seem to be working correctly for strapi. Hence. I did not use the migration files in this case.

File: src/api/idea-card/content-types/idea-card/lifecycles.js
Comment (oriyomibadmus) [inline src/api/idea-card/content-types/idea-card/lifecycles.js:67]: @royshunhung Could we add a matching duplicate check inside `beforeUpdate` (excluding the row being edited by id) so renames return the same friendly 400 instead of a raw DB error?

Comment (royshunhung) [reply src/api/idea-card/content-types/idea-card/lifecycles.js:67]: image

Confirmed latest change return HTTP400

**Review by oriyomibadmus — APPROVED** [review]

**Review by chungthuang — APPROVED** [review]

**Comment by dbradham** [comment]
Merged @royshunhung @AlinaU1
